datafusion/core/src/execution/context/mod.rs

```diff
@@ -2144,6 +2128,25 @@ impl FunctionRegistry for SessionState {
         })
     }
 
+    /// Add `analyzer_rule` to the `index` of the list of
```

Comment:
How would a user know what `index` to add I wonder 🤔  

Also, if the built in analyzer rules were changed, then any hard coded index might become incorrect

Instead of this API, what if we followed the model of  the `physical_optimizers()` APIL https://github.com/apache/arrow-datafusion/issues/9575

Specifically, we added a `SesssionContext::analyzer_rules()` API that returned all the existing rules? Using that and `with_analyzer_rules` the user could put the rules wherever they wanted

For example
```rust
let ctx = SessionContext::new();

let mut analyzer_rules = ctx.analyzer_rules();
// insert new rule at index 3
analyzer_rules.insert(3, new_rule);
// update rules in session context
let ctx = ctx.with_analyzer_rules(analyzer_rules);
```

(I think adding access to the analyzer rules makes sense even if we go with a different approach for rewrite)

Reply:
You are right, the example above is easier for the user to overwrite the rules. Let me add `analyzer_rules` to return the existing rules.

btw, is #9575 a correct link?

Reply:
I think i need to move `analyzer_rules` under FunctionRegistry, so it can be called in `register_all`

Reply:
> btw, is #9575 a correct link?

No sorry -- I meant to link to the `SessionCOntext::optimizers()` API. sorry about that